registration/models.py

```diff
@@ -376,3 +377,364 @@ def send_activation_email(self, site, request=None):
                 email_message.attach_alternative(message_html, 'text/html')
 
         email_message.send()
+
+
+class SupervisedRegistrationManager(RegistrationManager):
+
+    def activate_user(self, activation_key, site, get_profile=False):
+        """
+        Validate an activation key and activate the corresponding
+        registration profile object. The user is not yet activated
+        as the account is pending admin approval
+
+        If the key is valid and has not expired, return the ``True``
+        after activating.
+
+        If the key is not valid or has expired, return ``False``.
+
+        If the key is valid but the ``User`` is already active,
+        return ``True``.
+
+        If the key is valid but the ``User`` is inactive, return ``False``.
+
+        To prevent reactivation of an account which has been
+        deactivated by site administrators, ``SupervisedRegistrationProfile.activated``
+        is set to ``True`` after successful activation.
+
+        """
+        # Make sure the key we're trying conforms to the pattern of a
+        # SHA1 hash; if it doesn't, no point trying to look it up in
+        # the database.
+        if SHA1_RE.search(activation_key):
+            try:
+                profile = self.get(activation_key=activation_key)
+            except self.model.DoesNotExist:
+                # This is an actual activation failure as the activation
+                # key does not exist. It is *not* the scenario where an
+                # already activated User reuses an activation key.
+                return False
+
+            if profile.activated:
+                # The User has already activated and is trying to activate
+                # again. If the User is active, return the User. Else,
+                # return False as the User has not yet been approved by
+                # a site administrator or has been deactivated.
+                if profile.user.is_active:
+                    return True
+                else:
+                    return False
+
+            if not profile.activation_key_expired():
+                if not profile.user.is_active and not profile.activated:
+                    self.send_admin_approve_email(profile.user, site)
```

Comment:
I could, but I did not want to alter already existing code since I am implementing a new backend. But I guess it seems to make more sense this way.